pydantic/type_adapter.py

```diff
@@ -98,15 +115,33 @@ def _getattr_no_parents(obj: Any, attribute: str) -> Any:
         raise AttributeError(attribute)
 
 
+def _annotated_type(type_: Any) -> Any | None:
+    return get_args(type_)[0] if _typing_extra.is_annotated(type_) else None
+
+
 def _type_has_config(type_: Any) -> bool:
     """Returns whether the type has config."""
+    type_ = _annotated_type(type_) or type_
     try:
         return issubclass(type_, BaseModel) or is_dataclass(type_) or is_typeddict(type_)
     except TypeError:
         # type is not a class
         return False
 
 
+def _frame_depth(depth: int) -> Callable[[Callable[..., R]], Callable[..., R]]:
+    def wrapper(func: Callable[..., R]) -> Callable[..., R]:
+        @wraps(func)
+        def wrapped(self: TypeAdapter, *args: Any, **kwargs: Any) -> R:
+            # depth + 1 for the wrapper function
+            with self._with_frame_depth(depth + 1):
+                return func(self, *args, **kwargs)
+
+        return wrapped
+
+    return wrapper
+
+
```

Comment:
I think maybe it'd help to have a bit more documentation for this function and the frame depth function attached to the `TypeAdapter` class - it'd help me in my review as well!

Reply:
Done, added some comment to this.

FYI Im wondering is there a better way than the `_parent_depth` handling. What about requiring user to define a callback function that would resolve the forward refs from locals/globals? So replacing `_parent_depth` with something like `resolve_namespace: Callable[[str], type]` where the `str` is the name of the type to be resolved into the `type`. Then it wouldnt be as fragile as the parent depth handling.

This is ofcourse out of scope of this PR. It would require quite a bit of refactoring to the internals handling the `types_namespace` etc.

Reply:
I definitely wish we had a better way for users to specify/pass the namespace, but the reason for the parent depth thing was so that it would generally behave correctly without any extra work. At this point, I think it's probably not possible to get rid of `_parent_depth` unless we find a way to keep all existing code that currently relies on that from breaking. In v3 I think we could make a change to this if we felt it simplified things significantly or otherwise had (even indirect) import-time performance benefits

Reply:
Yeah no way to get rid of parent depth handling but some better (optional) way could be offered via configs to resolve the names. As working of the parent depth handling highly depends on the context on how model happens to be used